core/src/apps/cardano/sign_tx.py

```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
+    ) -> None:
         self.inputs = inputs
         self.outputs = outputs
         self.keychain = keychain
-        # attributes have to be always empty in current Cardano
-        self.attributes = {}
+        self.fee = fee
+        self.ttl = ttl
 
         self.network_name = KNOWN_PROTOCOL_MAGICS.get(protocol_magic, "Unknown")
         self.protocol_magic = protocol_magic
-        self.input_coins_sum = input_coins_sum
 
-    def _process_outputs(self):
-        change_addresses = []
-        change_derivation_paths = []
-        output_addresses = []
-        outgoing_coins = []
-        change_coins = []
+    def _build_inputs(self) -> list:
+        inputs_for_cbor = []
+        for input in self.inputs:
+            inputs_for_cbor.append([input.prev_hash, input.prev_index])
+
+        return inputs_for_cbor
+
+    def _build_outputs(self) -> list:
+        outputs_for_cbor = []
 
         for output in self.outputs:
+            amount = output.amount
             if output.address_n:
-                address, _ = derive_address_and_node(self.keychain, output.address_n)
-                change_addresses.append(address)
-                change_derivation_paths.append(output.address_n)
-                change_coins.append(output.amount)
+                address, _ = derive_address_and_node(
+                    self.keychain, output.address_n, self.protocol_magic
+                )
+                # set output address so it can later be shown to the user if needed
+                output.address = address
             else:
-                if output.address is None:
+                address = output.address
+                if address is None:
                     raise wire.ProcessError(
                         "Each output must have address or address_n field!"
                     )
-                if not is_safe_output_address(output.address):
+                if not is_safe_output_address(address):
                     raise wire.ProcessError("Invalid output address!")
 
-                outgoing_coins.append(output.amount)
-                output_addresses.append(output.address)
+            outputs_for_cbor.append([base58.decode(address), amount])
 
-        self.change_addresses = change_addresses
-        self.output_addresses = output_addresses
-        self.outgoing_coins = outgoing_coins
-        self.change_coins = change_coins
-        self.change_derivation_paths = change_derivation_paths
+        return outputs_for_cbor
 
-    def _build_witnesses(self, tx_aux_hash: str):
+    def _build_witnesses(self, tx_aux_hash: bytes) -> List[List[bytes]]:
         witnesses = []
         for input in self.inputs:
-            _, node = derive_address_and_node(self.keychain, input.address_n)
-            message = (
-                b"\x01" + cbor.encode(self.protocol_magic) + b"\x58\x20" + tx_aux_hash
-            )
+            node = self.keychain.derive(input.address_n)
+            message = b"\x58\x20" + tx_aux_hash
```

Comment:
do we know what are these magic numbers? isn't it some cbor?

Reply:
yes it's the CBOR prefix for a 32-byte byte sequence and this is probably a legacy from the way it was done in AdaLite: https://github.com/vacuumlabs/adalite/blob/bf73ee00cbd23f3648d80ba09e56873ca433479a/app/frontend/wallet/byron/cardano-wallet-secret-crypto-provider.ts#L89 - so it should be fine to do proper cbor encoding here

Reply:
Yes it is. Previously it was also done this way although now that I think about it, it was probably because protocol magic was somehow magically encoded cbor, right? 

I've changed it to `cbor.encode(tx_aux_hash)`